homeassistant/components/zimi/config_flow.py

```diff
@@ -0,0 +1,112 @@
+        description_placeholders: dict[str, str] = {}
+
+        if user_input is not None:
+            data: dict[str, Any] = {"title": "ZIMI Controller"}
+
+            try:
+                if not user_input[CONF_HOST]:
+                    try:
+                        description = await ControlPointDiscoveryService().discover()
+                    except ControlPointError:
+                        errors["base"] = "discovery_failure"
+                    data[CONF_HOST] = description.host
+                    data[CONF_PORT] = description.port
+                else:
+                    hostbyname = None
+                    data[CONF_HOST] = user_input[CONF_HOST]
+                    data[CONF_PORT] = user_input[CONF_PORT]
+                    try:
+                        hostbyname = socket.gethostbyname(data[CONF_HOST])
+                    except socket.gaierror as _:
+                        errors["base"] = "invalid_host"
+                    if hostbyname:
+                        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                        s.settimeout(10)
+                        try:
+                            s.connect((data[CONF_HOST], data[CONF_PORT]))
+                            s.close()
+                        except ConnectionRefusedError:
+                            errors["base"] = "connection_refused"
+                        except TimeoutError:
+                            errors["base"] = "timeout"
+                        except socket.gaierror:
+                            errors["base"] = "cannot_connect"
+
+                data[CONF_MAC] = format_mac(user_input[CONF_MAC])
+                if data[CONF_MAC] is user_input[CONF_MAC]:
+                    errors["base"] = "invalid_mac"
+
+                try:
+                    api = await async_connect_to_controller(
+                        data[CONF_HOST], data[CONF_PORT], fast=True
+                    )
+
+                    if api:
+                        if data[CONF_MAC] != format_mac(api.mac):
+                            msg = f"{data[CONF_MAC]} != {format_mac(api.mac)}"
+                            _LOGGER.error("Configured mac mismatch: %s", msg)
+                            errors["base"] = "mismatched_mac"
+                            description_placeholders["error_detail"] = msg
+
+                        api.disconnect()
+                    else:
+                        errors["base"] = "cannot_connect"
+
+                except ConfigEntryNotReady:
+                    errors["base"] = "cannot_connect"
+
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception during configuration steps")
+                errors["base"] = "unknown"
+
+            await self.async_set_unique_id(data[CONF_MAC])
+            self._abort_if_unique_id_configured()
+
+            if not errors:
+                return self.async_create_entry(title=data["title"], data=data)
```

Comment:
might as well set the title here directly instead of adding it to the dict first

Reply:
Done https://github.com/home-assistant/core/pull/129876/commits/a04741071d519c2e7ad7083c09bb9eaf6ca228ff